### DISCUSSION TITLE
[Feature request] Support for jest.mocks with independent modules

### DISCUSSION THREAD

**Discussion by rishitank** [Igorkowalski94/eslint-plugin-project-structure#18] [Ideas] [answered]
**Is your feature request related to a problem? Please describe.**
When defining import rules using independent modules, currently this doesn't apply to `jest.mock` and `jest.requireActual` imports in test files.

**Describe the solution you'd like**
I would like the same import rules that define who can import who, but with `jest.mock` and `jest.requireActual` imports too.

**Describe alternatives you've considered**
Currently it is a manual process to check import statements in test files.

**Additional context**

**Comment by rishitank** [comment]
Inspiration for this comes from `eslint-plugin-boundaries`: https://github.com/javierbrea/eslint-plugin-boundaries?tab=readme-ov-file#boundariesadditional-dependency-nodes

**Reply by Igorkowalski94** [reply]
Overall, I don't quite understand a scenario where this would make sense. 

For example, we have a certain component, and this component has its own import rules and cannot import anything beyond what is allowed for it. In the test file, we mock what the component uses, not something it doesn't use. So, if we additionally add a rule for jest.mock, it seems a bit redundant. I would prefer a rule that checks whether a particular mock is being used.


If, for example, we want to use `mockImplementation` or `mockReturnValue`, we still need to import the function, and then we will get an error if there's an issue.
```
import { validateRulesList } from "rules/folderStructure/helpers/validateChildren/helpers/validateRulesList";

jest.mock(
  "rules/folderStructure/helpers/validateChildren/helpers/validateRulesList",
  () => ({
    validateRulesList: jest.fn(),
  }),
);

const validateRulesListMock = jest.fn();

(validateRulesList as jest.Mock).mockImplementation(validateRulesListMock);
```

For example, if we make a typo, `jest` itself will provide information about it
```
    Cannot find module 'rules/folderStructure/helpers/validateChildren/help2ers/validateRulesList' from 'src/rules/folderStructure/helpers/validateChildren/validateChildren.test.ts'

      2 | import { validateRulesList } from "rules/folderStructure/helpers/validateChildren/helpers/validateRulesList";
      3 |
    > 4 | jest.mock(
        |      ^
      5 |   "rules/folderStructure/helpers/validateChildren/help2ers/validateRulesList",
      6 |   () => ({
      7 |     validateRulesList: jest.fn(),
```


Can you provide a scenario where it is actually not redundant?

**Reply by rishitank** [reply]
Yes a rule that checks whether a particular mock is being used would also be helpful. However there are situations where I have added a rule using `independent-modules` that errors on the import but not the corresponding `jest.mock` import. The import is valid in the sense that it imports a file that exists, but it should not be allowed according to my rules. For example:

- Say I have an import like this from a bad part of the codebase in a test file:
```
  import useSomeHook from '@components/my-component/hooks/use-some-hook';
```
I have setup my rules to make this path invalid, because technically you cannot access another component's private hooks. ESLint highlights this as an error which is correct.
- In my jest.mock I have:
```
  jest.mock('@components/my-component/hooks/use-some-hook', () => jest.fn());
```
This isn't highlighted by ESLint.
- Now say for example a developer were to expose that hook for whatever reason so that it could be accessed publicly:
```
import { useSomeHook } from '@components/my-component';
```
This may be seen as valid, however it could be that the developer might have forgotten to update the jest.mock, which according to ESLint has no issues, so without this validation the developer may not be motivated / remember to change it to:
```
jest.mock('@components/my-component', () => ({
  useSomeHook: jest.fn()
});
```
The mock would technically still work if it isn't refactored, but it would be using the wrong import path.

**Reply by Igorkowalski94** [reply]
But in that case, you're mocking a function that isn't used by the tested hook. So if you run `jest test`, the test will fail and inform you about it. And if the test passes, it means that the mock is not needed and should be removed. Then I would lean towards a rule that checks whether a particular mock is being used.

**Reply by Igorkowalski94** [reply, marked as answer]
A small update, `jest.mock`, `jest.requireActual ` import will be checked starting from version 2.7.0. The implementation itself was quite simple, so I decided to add it.
